src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/tools/monitoring/outputMonitor.ts

```diff
@@ -60,72 +65,210 @@ export class OutputMonitor extends Disposable implements IOutputMonitor {
 	constructor(
 		private readonly _execution: IExecution,
 		private readonly _pollFn: ((execution: IExecution, token: CancellationToken, taskService: ITaskService) => Promise<IPollingResult | undefined>) | undefined,
+		invocationContext: IToolInvocationContext,
+		token: CancellationToken,
+		command: string,
 		@ILanguageModelsService private readonly _languageModelsService: ILanguageModelsService,
 		@ITaskService private readonly _taskService: ITaskService,
 		@IChatService private readonly _chatService: IChatService,
 		@IChatWidgetService private readonly _chatWidgetService: IChatWidgetService
 	) {
 		super();
+		this.startMonitoring(command, invocationContext, token);
 	}
 
 	async startMonitoring(
 		command: string,
-		invocationContext: any,
+		invocationContext: IToolInvocationContext,
 		token: CancellationToken
-	): Promise<IPollingResult & { pollDurationMs: number }> {
+	): Promise<void> {
+		// Ensure this method always runs truly async so listeners can attach
+		await Promise.resolve();
 
 		const pollStartTime = Date.now();
+		let extended = false;
+		let autoReplyCount = 0;
+		let lastObservedLength = this._execution.getOutput().length;
+
+		while (!token.isCancellationRequested) {
+			const polled = await this._pollOnce(this._execution, extended, token, this._pollFn);
+
+			this._isIdle = polled.state === OutputMonitorState.Idle;
+			this._state = polled.state;
+
+			if (this._state === OutputMonitorState.Idle) {
+				this._onDidIdle.fire();
+			} else if (this._state === OutputMonitorState.Timeout) {
+				this._onDidTimeout.fire();
+			}
+
+			// If we timed out, optionally ask to keep waiting
+			if (this._state === OutputMonitorState.Timeout) {
```

Comment:
nit: switches are nicer when considering every state in a closed set of states like this imo